PINCache/PINOperationQueue.m

```diff
@@ -127,27 +135,79 @@ + (instancetype)sharedOperationQueue
 
 - (id <PINOperationReference>)addOperation:(dispatch_block_t)block
 {
-    return [self addOperation:block withPriority:PINOperationQueuePriorityDefault];
+  return [self addOperation:block withPriority:PINOperationQueuePriorityDefault];
 }
 
 - (id <PINOperationReference>)addOperation:(dispatch_block_t)block withPriority:(PINOperationQueuePriority)priority
 {
-  id <PINOperationReference> reference = [self nextOperationReference];
+  PINOperation *operation = [PINOperation operationWithBlock:^(id data) { block(); }
+                                                   reference:[self nextOperationReference]
+                                                    priority:priority
+                                                  identifier:nil
+                                                        data:nil
+                                                  completion:nil];
+  [self lock];
+    [self locked_addOperation:operation];
+  [self unlock];
   
-  NSMutableOrderedSet *queue = [self operationQueueWithPriority:priority];
+  [self scheduleNextOperations:NO];
   
-  PINOperation *operation = [PINOperation operationWithBlock:block reference:reference priority:priority];
+  return operation.reference;
+}
+
+- (id<PINOperationReference>)addOperation:(PINOperationBlock)block withPriority:(PINOperationQueuePriority)priority identifier:(NSString *)identifier data:(id)data dataCoallescingBlock:(PINOperationDataCoallescingBlock)dataCoallescingBlock completion:(dispatch_block_t)completion
+{
+  PINOperation *operation = nil;
+  BOOL isNewOperation = NO;
   
   [self lock];
-    dispatch_group_enter(_group);
-    [queue addObject:operation];
-    [_queuedOperations addObject:operation];
-    [_referenceToOperations setObject:operation forKey:reference];
+    if (identifier != nil && (operation = [_identifierToOperations objectForKey:identifier]) != nil) {
+      // There is an exisiting operation with the provided identifier, let's coallescing these operations
+      if (dataCoallescingBlock != nil) {
+        operation.data = dataCoallescingBlock(operation.data, data);
```

Comment:
I think we should probably unlock while calling this method in case someone tries to do something very expensive. We'd probably need to grab a reference to operation.data before unlocking.